src/main/java/duckling/Server.java

```diff
@@ -1,11 +1,101 @@
 package duckling;
 
+import duckling.requests.Request;
+import duckling.requests.RequestBuilder;
+import duckling.responders.Responders;
+
+import java.io.*;
+import java.net.*;
+import java.util.concurrent.*;
+
 public class Server {
+    public static final String CLRF = "\r\n";
+
     public int port;
+    public String root;
+
+    private boolean shuttingDown = false;
+    private InetSocketAddress address;
+    private ServerSocket connection;
+    private ExecutorService pool;
+
+    public Server(int port, String root) throws IOException {
+        this(
+            port,
+            root,
+            new ServerSocket(),
+            Executors.newCachedThreadPool()
+        );
+    }
 
-    public Server(int port) {
+    public Server(
+        int port,
+        String root,
+        ServerSocket connection,
+        ExecutorService pool
+    ) throws IOException {
         this.port = port;
+        this.root = root;
+
+        this.address = new InetSocketAddress(this.port);
+        this.connection = connection;
+        this.pool = pool;
+    }
+
+    public boolean isPoolClosed() {
+        return this.pool.isShutdown();
+    }
+
+    public void onBegin() throws IOException {
+        this.connection.bind(this.address);
+    }
+
+    public void onRequest() throws IOException {
+        Socket client = this.connection.accept();
+        Runnable handler = () -> {
+            try {
+                RequestBuilder builder = new RequestBuilder(
+                    this.root,
+                    client.getInputStream()
+                );
+                Request request = builder.build();
+                Responders responders = new Responders(
+                    request,
+                    client.getOutputStream()
+                );
+
+                builder.printRawRequest();
+                responders.getResponder().respond();
+
+                client.close();
+            } catch (IOException exception) {
+                exception.printStackTrace();
+            }
+        };
+
+        this.pool.execute(handler);
+    }
+
+    public void onShutdown() {
+        try {
+            this.pool.shutdown();
+            this.connection.close();
+            this.shuttingDown = true;
```

Comment:
If the loop below checks this `shuttingDown` variable concurrently as the `pool.shutdown()` and `connection.close()` are happening, it seems like the request processing would fail due to connections being closed out from under it. If `shuttingDown` was set first, it closes the gap a bit but a similar race condition still exists (we could whiteboard it to look at the interleaving). I think only a lock (`synchronized` or other) would prevent connections being closed out from under consumers.

Of course, this would only happen during shutdown, so it might seem OK either way from a user perspective (and certainly for Fitnesse tests). Just some more food for thought on the complexity of multithreading.